crates/ika-core/src/dwallet_mpc/crytographic_computation/mpc_computations/presign.rs

```diff
@@ -4,66 +4,306 @@
+            }
+            VersionedDwalletDKGSecondRoundPublicOutput::V2(dkg_output) => Self::try_new_v2(
+                protocol,
+                versioned_network_encryption_key_public_data,
+                Some(dkg_output),
+            ),
+        }
+    }
+    pub(crate) fn try_new_v1(
+        versioned_network_encryption_key_public_data: &VersionedNetworkEncryptionKeyPublicData,
+        dwallet_public_output: MPCPublicOutput,
+    ) -> DwalletMPCResult<Self> {
+        let decentralized_party_dkg_output = bcs::from_bytes::<
+            <Secp256K1ECDSAProtocol as dkg::Protocol>::DecentralizedPartyTargetedDKGOutput,
+        >(&dwallet_public_output)?
+        .into();
 
-        let pub_input = Self::PublicInput {
-            protocol_public_parameters,
-            dkg_output: decentralized_dkg_output,
+        let protocol_public_parameters =
+            versioned_network_encryption_key_public_data.secp256k1_protocol_public_parameters();
+
+        let public_input: <PresignParty<Secp256K1ECDSAProtocol> as mpc::Party>::PublicInput =
+            (protocol_public_parameters, decentralized_party_dkg_output).into();
+        Ok(PresignPublicInputByProtocol::Secp256k1ECDSA(public_input))
+    }
+
+    pub(crate) fn try_new_v2(
+        protocol: DWalletSignatureScheme,
+        versioned_network_encryption_key_public_data: &VersionedNetworkEncryptionKeyPublicData,
+        dwallet_public_output: Option<MPCPublicOutput>,
+    ) -> DwalletMPCResult<Self> {
+        let input = match protocol {
+            DWalletSignatureScheme::ECDSASecp256k1 => {
+                let protocol_public_parameters = versioned_network_encryption_key_public_data
+                    .secp256k1_protocol_public_parameters();
+
+                let deserialized_dwallet_public_output =
+                    dwallet_public_output.and_then(|bytes| bcs::from_bytes(&bytes).ok());
+                let pub_input =
+                    <PresignParty<Secp256K1ECDSAProtocol> as mpc::Party>::PublicInput::from((
+                        protocol_public_parameters,
+                        deserialized_dwallet_public_output,
+                    ));
+                PresignPublicInputByProtocol::Secp256k1ECDSA(pub_input)
+            }
+            DWalletSignatureScheme::SchnorrkelSubstrate => {
+                let protocol_public_parameters = versioned_network_encryption_key_public_data
+                    .ristretto_protocol_public_parameters()?;
+                let deserialized_dwallet_public_output: Option<
+                    <RistrettoAsyncDKGProtocol as dkg::Protocol>::DecentralizedPartyDKGOutput,
+                > = dwallet_public_output.and_then(|bytes| bcs::from_bytes(&bytes).ok());
+                let pub_input =
+                    <PresignParty<RistrettoSchnorrkelSubstrateProtocol> as mpc::Party>::PublicInput::from((protocol_public_parameters, deserialized_dwallet_public_output));
```

Comment:
deserialization is good, but how is the code compiling, if the `from()` expects `Option<Self::DecentralizedPartyTargetedDKGOutput>`?

Reply:
Also the `and_then` and `.ok()` is bad here, we should return an error and reject session in case the deserialization failed